### ISSUE TITLE
[QST] RNG Lib

### ISSUE THREAD

**Issue by mnicely** [HazyResearch/flash-attention#56]
Hi, IIUC, you're using a PyTorch RNG for the device. Have you considered cuRAND's device side API?

**Comment by tridao** [comment]
Hi Matthew.
I started with cuRAND and it's a bit slower, since for Philox it does 10 rounds while we do 7 rounds. Apex also uses [7-round Philox](https://github.com/NVIDIA/apex/blob/master/apex/contrib/csrc/multihead_attn/philox.cuh) for speed.

**Comment by mnicely** [comment]
Okay, thank you for the feedback. Let me see what we can do to improve performance and get back to you. Can you tell me what data types and distributions you need most?

**Comment by tridao** [comment]
Thanks Matthew, that would be great!
We use it for dropout, so just `uniform` distribution on [0, 1) and data type `float`, with the Philox RNG. I believe Apex uses it for the same setting.
If one does dropout as a separate kernel (e.g., Pytorch's nn.Dropout) then the RNG speed doesn't matter that much since the operation is memory-bandwidth bound. But as one starts to fuse operations (as we do in FlashAttention and in Apex's FMHA), the RNG speed start to matter.

**Comment by tridao** [comment]
Btw Apex's implementation has [inline assembly](https://github.com/NVIDIA/apex/pull/1265) to speed up Philox RNG, idk that's already in cuRAND.
